### DISCUSSION TITLE
Lost calendar, address, tasks after server crash

### DISCUSSION THREAD

**Discussion by charleslcso** [Kozea/Radicale#1574] [Q&A server] [answered]
So happen that our Proxmox LXC containers all shutdown unexpectedly. Bad stuff.

Then realise all users' Radicale data is gone: addresses, calendar events, tasks/reminders.

I'll need to backup Radicale regularly. Also, how should this kind of event be handled?

**Comment by pbiering** [comment]
non-transient data should never stored inside containers but outside on stable and backup-ed file system...

usually, all files in the collections-root main folder should be backed-up, simple full backup e.g.

`tar czf /path/to/radicale-collection-root-backup-$(date +'%Y%m%d-%H%M%S').tar.gz /var/lib/radicale/collections/collection-root/`

one can think about excluding the cache directories,

frequency strongly depending on the change rate

also one can use a more sophisticated tools like `amanda` or `duplicity` to run incremental backups and able to have a history in case a particular file only need to be restored

**Reply by charleslcso** [reply]
Hello @pbiering , thanks for your reply. 

[This](https://floatingoctothorpe.uk/2017/automatic-radicale-backups-with-git.html) seems very interesting and workable for me. Do you have any comment for v3?

**Reply by pbiering** [reply]
this is related to v3, storage layout is file based and documented here https://github.com/Kozea/Radicale/wiki/Collection-Storage

**Reply by charleslcso** [reply]
Hi Peter,

When I come and think of it, I had a container shutdown uncleanly. 

All user's data is gone. At least the user sees no data at all.

Saving the directory content in another stable device shouldn't help in this case, I believe, because after the reboot:

1. all files are corrupted in the container or
2. all files vanish because radicale didn't stop properly

Stable drive/devices shouldn't have any benefit in both situations. 

Please correct me if I'm wrong.

**Reply by pbiering** [reply]
this strongly depends how far the (hopefully journaling) file system was abstracted.

per se a container has no persistent file system at all internally, only in-memory

if a container requires persistent file system, this has to be mounted from outside into the container

if the "outside" crashes itself it's strongly depending of which kind it was...

- a virtual machine based on a disk image
or
- a physical system having disks (RAID)

-> as long as the (virtual) disk is not damaged, the file system should be able to be recover

And _radicale_ is not vanishing files on stop, can it be that you run _radicale_ in a container without persistent storage, e.g. using only _tempfs_ in memory instead of a persistent file system mounted into the container (from the host or via NFS, iSCSI  or something else)?

**Reply by charleslcso** [reply]
Hi @pbiering 

The container running radicale is using ext4 fs, from host. Yes, there are tmpfs in the container, but they are not used by radicale.

Using the default radicale parameters, collections are stored in /root/.var/lib/radicale/collections/collection-root/<users>/, which means /root is not in tmpfs.

**Reply by pbiering** [reply]
_radicale_ is not deleting files on startup and _ext4_ filesystem usually has a journal -> the files should be even after a container or host crash visible on the host file system, except some cleanup magic by used infrastructure either on crash/shutdown or (re)start wipes such mapped directory. Backup should be also done on host level and pushed to or pulled from another host.

**Reply by charleslcso** [reply, marked as answer]
> non-transient data should never stored inside containers but outside on stable and backup-ed file system...
> 
> usually, all files in the collections-root main folder should be backed-up, simple full backup e.g.
> 
> `tar czf /path/to/radicale-collection-root-backup-$(date +'%Y%m%d-%H%M%S').tar.gz /var/lib/radicale/collections/collection-root/`
> 
> one can think about excluding the cache directories,
> 
> frequency strongly depending on the change rate
> 
> also one can use a more sophisticated tools like `amanda` or `duplicity` to run incremental backups and able to have a history in case a particular file only need to be restored

Peter,

Shit, it happens again. I must restart the server, and now all my Reminders and Calendars are mostly gone! Except there are 1 calendar and 1 addressbook in Web UI.

All other calendars have vanished.

I made frequent backups on another server, and I restored it in /.var/lib/radicale/collections/collection-root/ (this is where radicale stores files in Debian), but still, in Web UI, nothing seems to have restored.

By checking the directory listing of the backup, content size wise it should be good. 

What to do what to do?

**Reply by pbiering** [reply]
Have you verified permission of files after restore?

Check also whether the structure was proper restored
https://github.com/Kozea/Radicale/wiki/Collection-Storage

Also run storage verification (best in debug mode)
https://github.com/Kozea/Radicale/wiki/Server-Diagnostics---Troubleshooting

In addition I strongly recommend mounting a persistent storage from host or network into the container or install software itself on a host or inside a VM (with persistent disk image...) to avoid such issue again.

**Reply by charleslcso** [reply]
> Have you verified permission of files after restore?
> 
> Check also whether the structure was proper restored https://github.com/Kozea/Radicale/wiki/Collection-Storage
> 
> Also run storage verification (best in debug mode) https://github.com/Kozea/Radicale/wiki/Server-Diagnostics---Troubleshooting
> 
> In addition I strongly recommend mounting a persistent storage from host or network into the container or install software itself on a host or inside a VM (with persistent disk image...) to avoid such issue again.

Radicale is installed in a VM with persistent disk image. I have no idea why it happens. It was a proper shutdown.

What's in the /root/.var/..../collection, with the dot in front of var?

**Reply by pbiering** [reply]
the `.` in front of a directory name is turning the directory into "hidden" - used in Unix since many many years. There are also such other directories in parallel in an user's home directory.

Running radicale interactive and especially as root user is not recommended for any productive usage.

For your skill set I strongly recommend using packaged (DEB/RPM) radicale installation for the operating system you are using.
They are also containing scripts for running as service and usually store data under /var/lib/.

If your current OS unfortunately not up-to-date versions either trigger maintainer to update or switch the OS e.g. to Enterprise Linux (e.g. Alma - usually 5 year lifecycle) or Fedora Linux (if 1,5 yearlifecycle is acceptable) as they have up-to-date support.
